pkg/deployment/deployment.go

```diff
@@ -0,0 +1,379 @@
+	RecoverySize  = 0
+
+	SystemLabel      = "SYSTEM"
+	SystemMnt        = "/"
+	AllAvailableSize = -1
+
+	Unknown = "unknown"
+)
+
+type PartFunction int
+
+const (
+	EFI PartFunction = iota + 1
+	System
+	Recovery
+	Data
+)
+
+type FileSystem int
+
+const (
+	Btrfs FileSystem = iota + 1
+	Ext2
+	Ext4
+	XFS
+	VFat
+)
+
+func ParseFileSystem(f string) (FileSystem, error) {
+	switch f {
+	case "btrfs":
+		return Btrfs, nil
+	case "ext2":
+		return Ext2, nil
+	case "ext4":
+		return Ext4, nil
+	case "xfs":
+		return XFS, nil
+	case "vfat":
+		return VFat, nil
+	default:
+		return FileSystem(0), fmt.Errorf("filesystem not supported: %s", f)
+	}
+}
+
+func (f FileSystem) String() string {
+	switch f {
+	case Btrfs:
+		return "btrfs"
+	case Ext2:
+		return "ext2"
+	case Ext4:
+		return "ext4"
+	case XFS:
+		return "xfs"
+	case VFat:
+		return "vfat"
+	default:
+		return Unknown
+	}
+}
+
+func (f FileSystem) MarshalJSON() ([]byte, error) {
+	return json.Marshal(f.String())
+}
+
+func (f *FileSystem) UnmarshalJSON(data []byte) (err error) {
+	var function string
+	if err = json.Unmarshal(data, &function); err != nil {
+		return err
+	}
+
+	*f, err = ParseFileSystem(function)
+	return err
+}
+
+func ParseFunction(function string) (PartFunction, error) {
+	switch function {
+	case "efi":
+		return EFI, nil
+	case "system":
+		return System, nil
+	case "recovery":
+		return Recovery, nil
+	case "data":
+		return Data, nil
+	default:
+		return PartFunction(0), fmt.Errorf("unknown partition function: %s", function)
+	}
+}
+
+func (p PartFunction) String() string {
+	switch p {
+	case EFI:
+		return "efi"
+	case System:
+		return "system"
+	case Recovery:
+		return "recovery"
+	case Data:
+		return "data"
+	default:
+		return Unknown
+	}
+}
+
+func (p PartFunction) MarshalJSON() ([]byte, error) {
+	return json.Marshal(p.String())
+}
+
+func (p *PartFunction) UnmarshalJSON(data []byte) (err error) {
+	var function string
+	if err = json.Unmarshal(data, &function); err != nil {
+		return err
+	}
+
+	*p, err = ParseFunction(function)
+	return err
+}
+
+type RWVolume struct {
+	Path          string   `json:"path"`
+	Snapshotted   bool     `json:"snapshotted,omitempty"`
+	NoCopyOnWrite bool     `json:"noCopyOnWrite,omitempty"`
+	MountOpts     []string `json:"mountOpts,omitempty"`
+}
+
+type Partition struct {
+	Label       string       `json:"label,omitempty"`
+	FileSystem  FileSystem   `json:"fileSystem,omitempty"`
+	Size        int64        `json:"size,omitempty"`
+	Function    PartFunction `json:"function"`
+	StartSector uint         `json:"startSector,omitempty"`
+	MountPoint  string       `json:"mountPoint,omitempty"`
+	MountOpts   string       `json:"mountOpts,omitempty"`
+	RWVolumes   []RWVolume   `json:"rwVolumes,omitempty"`
+}
+
+type Disk struct {
+	Device     string       `json:"device,omitempty"`
+	Partitions []*Partition `json:"partitions"`
+	SectorSize uint         `json:"sectorSize,omitempty"`
+}
+
+type Deployment struct {
```

Comment:
I don't see such term in the existing `elemental-toolkit` codebase. Could you please explain why it is picked here?

P.S. Note that I'm not against calling it a "Deployment", I'm only looking to understand the idea behind it.

Reply:
I am not really good at naming. I used `Deployment` because it describes the OS deployment on that node and I did not want to call it `Install` as I expect to use it for `upgrade`, `build` and  `reset` (if ever consider it). My idea is that after the installation this is serialized and stored in `/etc/elemental/deployment.yaml` (or a similar file) so this can be easily loaded in `upgrade` process to get the current system state, preventing the need of running heuristics to figure out partition labels, UUIDs, RW paths, etc.

Reply:
Maybe `OSDeployment` in that case?